File: internal/threads/threads.go
// Package threads persists named chat conversations under
// <project>/.documind/chats/<name>.json. Shape mirrors the Python threads module:
// a thread has a name, timestamps, and an ordered list of user/assistant
// messages. A windowed slice of recent turns feeds the answer prompt (never the
// classify/clarify prompts, which stay single-purpose).
package threads

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/Henildiyora/DocuMind/internal/config"
	"github.com/Henildiyora/DocuMind/internal/ollama"
)

// DefaultThread is the thread name used by one-shot `documind ask`.
const DefaultThread = "default"

var nameRe = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)

// Message is one turn in a conversation.
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
	TS      string `json:"ts,omitempty"`
}

// Thread is a named, persisted conversation.
type Thread struct {
	Name      string    `json:"name"`
	CreatedAt string    `json:"created_at"`
	UpdatedAt string    `json:"updated_at"`
	Messages  []Message `json:"messages"`
}

func nowUTC() string {
	return time.Now().UTC().Truncate(time.Second).Format(time.RFC3339)
}

// SanitizeName validates and returns a safe thread name.
func SanitizeName(name string) (string, error) {
	cleaned := strings.TrimSpace(name)
	if cleaned == "" || !nameRe.MatchString(cleaned) {
		return "", fmt.Errorf("invalid thread name %q. Use letters, digits, _ or - only", name)
	}
	return cleaned, nil
}

// ChatsDir returns <project>/.documind/chats.
func ChatsDir(projectRoot string, cfg config.Config) string {
	return filepath.Join(cfg.IndexDirFor(projectRoot), "chats")
}

func threadPath(projectRoot string, name string, cfg config.Config) (string, error) {
	safe, err := SanitizeName(name)
	if err != nil {
		return "", err
	}
	return filepath.Join(ChatsDir(projectRoot, cfg), safe+".json"), nil
}

// Load reads a thread from disk, or returns a fresh in-memory one if absent.
func Load(projectRoot, name string, cfg config.Config) (*Thread, error) {
	safe, err := SanitizeName(name)
	if err != nil {
		return nil, err
	}
	path, _ := threadPath(projectRoot, safe, cfg)
	data, err := os.ReadFile(path)
	if err != nil {
		return &Thread{Name: safe, CreatedAt: nowUTC(), UpdatedAt: nowUTC()}, nil
	}
	var t Thread
	if err := json.Unmarshal(data, &t); err != nil {
		return &Thread{Name: safe, CreatedAt: nowUTC(), UpdatedAt: nowUTC()}, nil
	}
	t.Name = safe
	if t.CreatedAt == "" {
		t.CreatedAt = nowUTC()
	}
	return &t, nil
}

// Save persists a thread, updating its UpdatedAt timestamp.
func Save(projectRoot string, t *Thread, cfg config.Config) error {
	path, err := threadPath(projectRoot, t.Name, cfg)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	t.UpdatedAt = nowUTC()
	data, err := json.MarshalIndent(t, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, append(data, '\n'), 0o644)
}

// AppendTurn appends a user/assistant turn to a thread and saves it.
func AppendTurn(projectRoot, name, user, assistant string, cfg config.Config) error {
	t, err := Load(projectRoot, name, cfg)
	if err != nil {
		return err
	}
	ts := nowUTC()
	t.Messages = append(t.Messages,
		Message{Role: "user", Content: user, TS: ts},
		Message{Role: "assistant", Content: assistant, TS: ts},
	)
	return Save(projectRoot, t, cfg)
}

// List returns all threads sorted by UpdatedAt descending.
func List(projectRoot string, cfg config.Config) ([]*Thread, error) {
	dir := ChatsDir(projectRoot, cfg)
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, nil // no chats dir yet: not an error
	}
	var out []*Thread
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".json") {
			continue
		}
		name := strings.TrimSuffix(e.Name(), ".json")
		t, err := Load(projectRoot, name, cfg)
		if err != nil {
			continue
		}
		out = append(out, t)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].UpdatedAt > out[j].UpdatedAt })
	return out, nil
}

// Ensure loads or creates and persists an empty thread.
func Ensure(projectRoot, name string, cfg config.Config) (*Thread, error) {
	t, err := Load(projectRoot, name, cfg)
	if err != nil {
		return nil, err
	}
	path, _ := threadPath(projectRoot, t.Name, cfg)
	if _, statErr := os.Stat(path); statErr != nil {
		if err := Save(projectRoot, t, cfg); err != nil {
			return nil, err
		}
	}
	return t, nil
}

// Rename renames a thread file and updates its name.
func Rename(projectRoot, oldName, newName string, cfg config.Config) (*Thread, error) {
	oldSafe, err := SanitizeName(oldName)
	if err != nil {
		return nil, err
	}
	newSafe, err := SanitizeName(newName)
	if err != nil {
		return nil, err
	}
	if oldSafe == newSafe {
		return Load(projectRoot, oldSafe, cfg)
	}
	srcPath, _ := threadPath(projectRoot, oldSafe, cfg)
	dstPath, _ := threadPath(projectRoot, newSafe, cfg)
	if _, err := os.Stat(srcPath); err != nil {
		return nil, fmt.Errorf("thread %q does not exist", oldSafe)
	}
	if _, err := os.Stat(dstPath); err == nil {
		return nil, fmt.Errorf("thread %q already exists", newSafe)
	}
	t, err := Load(projectRoot, oldSafe, cfg)
	if err != nil {
		return nil, err
	}
	t.Name = newSafe
	if err := Save(projectRoot, t, cfg); err != nil {
		return nil, err
	}
	_ = os.Remove(srcPath)
	return t, nil
}

// Delete removes a thread file if present.
func Delete(projectRoot, name string, cfg config.Config) error {
	path, err := threadPath(projectRoot, name, cfg)
	if err != nil {
		return err
	}
	if _, statErr := os.Stat(path); statErr == nil {
		return os.Remove(path)
	}
	return nil
}

// Preview returns a short last-message preview for `/threads`.
func (t *Thread) Preview(limit int) string {
	if len(t.Messages) == 0 {
		return "(empty)"
	}
	text := strings.ReplaceAll(strings.TrimSpace(t.Messages[len(t.Messages)-1].Content), "\n", " ")
	if len(text) > limit {
		return text[:limit-1] + "…"
	}
	if text == "" {
		return "(empty)"
	}
	return text
}

// HistoryWindow returns the last `turns` user+assistant pairs as chat messages
// for prompt memory. Truncates to the window; never includes timestamps.
func (t *Thread) HistoryWindow(turns int) []ollama.Message {
	n := turns * 2
	if n <= 0 || len(t.Messages) == 0 {
		return nil
	}
	start := len(t.Messages) - n
	if start < 0 {
		start = 0
	}
	var out []ollama.Message
	for _, m := range t.Messages[start:] {
		out = append(out, ollama.Message{Role: m.Role, Content: m.Content})
	}
	return out
}

File: src/documind/threads.py
"""Persistent named chat threads under ``.documind/chats/``."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path

from .config import Config, index_dir_for, load_config

_NAME_RE = re.compile(r"^[a-zA-Z0-9_-]+$")
DEFAULT_THREAD = "default"


def _utc_now() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat()


def sanitize_thread_name(name: str) -> str:
    """Validate and return a safe thread name."""
    cleaned = (name or "").strip()
    if not cleaned or not _NAME_RE.match(cleaned):
        raise ValueError(
            f"Invalid thread name {name!r}. Use letters, digits, _ or - only."
        )
    return cleaned


def chats_dir(project_root: Path, cfg: Config | None = None) -> Path:
    """Return ``<project>/.documind/chats``."""
    cfg = cfg or load_config()
    return index_dir_for(project_root, cfg) / "chats"


@dataclass
class Thread:
    """A named conversation stored as JSON."""

    name: str
    created_at: str = field(default_factory=_utc_now)
    updated_at: str = field(default_factory=_utc_now)
    messages: list[dict] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "messages": self.messages,
        }

    @classmethod
    def from_dict(cls, data: dict) -> Thread:
        return cls(
            name=str(data.get("name", DEFAULT_THREAD)),
            created_at=str(data.get("created_at") or _utc_now()),
            updated_at=str(data.get("updated_at") or _utc_now()),
            messages=list(data.get("messages") or []),
        )

    def preview(self, limit: int = 60) -> str:
        """Last message content preview for ``/threads``."""
        if not self.messages:
            return "(empty)"
        text = str(self.messages[-1].get("content") or "").strip().replace("\n", " ")
        if len(text) > limit:
            return text[: limit - 1] + "…"
        return text or "(empty)"

    def history_window(self, turns: int) -> list[dict]:
        """Return the last ``turns`` user+assistant pairs as Ollama messages."""
        n = max(0, int(turns)) * 2
        if n <= 0:
            return []
        window = self.messages[-n:]
        return [{"role": m["role"], "content": m["content"]} for m in window]


def _thread_path(project_root: Path, name: str, cfg: Config | None = None) -> Path:
    safe = sanitize_thread_name(name)
    return chats_dir(project_root, cfg) / f"{safe}.json"


def load_thread(
    project_root: Path,
    name: str = DEFAULT_THREAD,
    cfg: Config | None = None,
) -> Thread:
    """Load a thread from disk, or create an empty in-memory one."""
    path = _thread_path(project_root, name, cfg)
    if not path.exists():
        return Thread(name=sanitize_thread_name(name))
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        thread = Thread.from_dict(data)
        thread.name = sanitize_thread_name(name)
        return thread
    except (OSError, json.JSONDecodeError, ValueError):
        return Thread(name=sanitize_thread_name(name))


def save_thread(
    project_root: Path,
    thread: Thread,
    cfg: Config | None = None,
) -> Path:
    """Persist ``thread`` to ``.documind/chats/<name>.json``."""
    path = _thread_path(project_root, thread.name, cfg)
    path.parent.mkdir(parents=True, exist_ok=True)
    thread.updated_at = _utc_now()
    path.write_text(json.dumps(thread.to_dict(), indent=2) + "\n", encoding="utf-8")
    return path


def append_turn(
    project_root: Path,
    name: str,
    user: str,
    assistant: str,
    cfg: Config | None = None,
) -> Thread:
    """Append a user/assistant turn and save."""
    thread = load_thread(project_root, name, cfg)
    now = _utc_now()
    thread.messages.append({"role": "user", "content": user, "ts": now})
    thread.messages.append({"role": "assistant", "content": assistant, "ts": now})
    save_thread(project_root, thread, cfg)
    return thread


def list_threads(project_root: Path, cfg: Config | None = None) -> list[Thread]:
    """Return all threads sorted by ``updated_at`` descending."""
    root = chats_dir(project_root, cfg)
    if not root.exists():
        return []
    threads: list[Thread] = []
    for path in sorted(root.glob("*.json")):
        try:
            threads.append(load_thread(project_root, path.stem, cfg))
        except ValueError:
            continue
    threads.sort(key=lambda t: t.updated_at, reverse=True)
    return threads


def rename_thread(
    project_root: Path,
    old: str,
    new: str,
    cfg: Config | None = None,
) -> Thread:
    """Rename a thread file and update its name field."""
    old_safe = sanitize_thread_name(old)
    new_safe = sanitize_thread_name(new)
    if old_safe == new_safe:
        return load_thread(project_root, old_safe, cfg)
    src = _thread_path(project_root, old_safe, cfg)
    dst = _thread_path(project_root, new_safe, cfg)
    if not src.exists():
        raise FileNotFoundError(f"Thread {old_safe!r} does not exist.")
    if dst.exists():
        raise FileExistsError(f"Thread {new_safe!r} already exists.")
    thread = load_thread(project_root, old_safe, cfg)
    thread.name = new_safe
    save_thread(project_root, thread, cfg)
    src.unlink(missing_ok=True)
    return thread


def delete_thread(
    project_root: Path,
    name: str,
    cfg: Config | None = None,
) -> None:
    """Delete a thread file if present."""
    path = _thread_path(project_root, name, cfg)
    if path.exists():
        path.unlink()


def ensure_thread(
    project_root: Path,
    name: str,
    cfg: Config | None = None,
) -> Thread:
    """Load or create and persist an empty thread."""
    thread = load_thread(project_root, name, cfg)
    path = _thread_path(project_root, thread.name, cfg)
    if not path.exists():
        save_thread(project_root, thread, cfg)
    return thread
